src/core/qgsvectorfilewriter.cpp

```diff
@@ -2842,67 +2839,83 @@ QList< QgsVectorFileWriter::FilterFormatDetails > QgsVectorFileWriter::supported
         if ( filterString.isEmpty() )
           continue;
 
+        MetaData metadata;
+        QStringList globs;
+        if ( driverMetadata( drvName, metadata ) && !metadata.glob.isEmpty() )
+        {
+          globs = metadata.glob.toLower().split( ' ' );
+        }
+
         FilterFormatDetails details;
         details.driverName = drvName;
         details.filterString = filterString;
-
-        if ( options & SortRecommended )
-        {
-          if ( drvName == QLatin1String( "ESRI Shapefile" ) )
-          {
-            shapeFormat = details;
-            continue;
-          }
-          else if ( drvName == QLatin1String( "GPKG" ) )
-          {
-            gpkgFormat = details;
-            continue;
-          }
-        }
+        details.globs = globs;
 
         results << details;
       }
     }
   }
 
-  std::sort( results.begin(), results.end(), []( const FilterFormatDetails & a, const FilterFormatDetails & b ) -> bool
-  {
-    return a.driverName < b.driverName;
-  } );
-
-  if ( options & SortRecommended )
+  std::sort( results.begin(), results.end(), [options]( const FilterFormatDetails & a, const FilterFormatDetails & b ) -> bool
   {
-    if ( !shapeFormat.filterString.isEmpty() )
-    {
-      results.insert( 0, shapeFormat );
-    }
-    if ( !gpkgFormat.filterString.isEmpty() )
+    if ( options & SortRecommended )
     {
-      results.insert( 0, gpkgFormat );
+      if ( a.driverName == QLatin1String( "GPKG" ) )
+        return true; // Make https://twitter.com/shapefiIe a sad little fellow
+      else if ( b.driverName == QLatin1String( "GPKG" ) )
+        return false;
+      else if ( a.driverName == QLatin1String( "ESRI Shapefile" ) )
```

Comment:
Why not using the alphabetical position for shp?

Reply:
Because it's still the most commonly used format for spatial data, so it needs to be readily accessible and not buried in a list amongst the rarer formats.

Reply:
Oh, I thought you were sending that to the bottom, but now I see it's just number 2. Ok for me :+1:

Reply:
Yeah this doesn't change the 3.0 behaviour, it just fixes sorting of the other formats.

So ok to merge?